exercises/connect/example.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * "Player O" plays from top to bottom, "Player X" plays from left to right.
+ * @param {string[]} initBoard - The starting state of the board
+ * @returns {Object} - The board business object
+ */
+var Board = function (initBoard) {
+  this.board = initBoard.slice();
+};
+
+Board.prototype.winner = function () {
+  var players = ['X', 'O'];
+  for (var i = 0; i < players.length; i++) {
+    if (this.checkWin(players[i])) {
+      return players[i];
+    }
+  }
+  return '';
+};
+
+Board.prototype.checkWin = function (player) {
+  var positions = this.startPositions(player);
+  for (var i = 0; i < positions.length; i++) {
+    if (this.search(positions[i], player, [])) {
+      return true;
+    }
+  }
+  return false;
+};
+
+Board.prototype.search = function (pos, XorO, isChecked) {
+  var boardInstance = this;
+  if (!this.matches(pos, XorO)) {
+    return false;
+  }
+  if (this.winningSpot(pos, XorO)) {
+    return true;
+  }
+  var checked = isChecked.slice(0);
+  checked.push(pos);
+
+  var matches = this.neighbors(pos).filter(function (cell) {
+    return boardInstance.matches(cell, XorO) &&
+      checked.filter(function (spot) { return spot.x === cell.x && spot.y === cell.y; }).length === 0;
```

Comment:
What about chainning a couple of filters?

```javascript
checked
    .filter(function (spot) { return spot.x === cell.x; })
    .filter(function (spot) { return spot.y === cell.y; })
    .length === 0;
```

Does it the code clearer?

Reply:
I agree that it reads better.  It also means every element potentially gets evaluated twice.  I don't think this is a big deal in this exercise, but maybe as a best practice it would be better to extract both check into one named function.

Reply:
It sounds to me like *premature optimization* 😝 

I usually favor readibility over performance (when performance is not an issue).

Even in real projects, if you're looping 100 items, it's almost the same looping them once than twice. Anyway, a well named function is much better than lot of things 😄